### PR TITLE
MTD Validation: SIM clusters

### PR DESCRIPTION
**Pull request by casarsa**
#### PR description:

This PR builds the SIM clusters associated to the reconstructed BTL and ETL clusters and fills new monitoring histograms with:
- time, energy and position resolutions for BTL and ETL clusters;
- cluster time pulls vs energy and eta;
- Y vs X of the cluster local coordinates.

(@fabiocos , @parbol , @gsorrentino18 )

#### PR validation:

The new code has been validated with the TTbar WFs 34634.0 and 34834.0 with CMSSW_12_0_X_2021-06-21-1100.

--------------------------------------------------------------------------------------

**current code**

step 3 (WF 34634.0)

MemoryReport> Peak virtual size 6372.39 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 6372.34 deltaVsize = 0 rss = 3253.47 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 6372.39 deltaVsize = 0.046875 rss = 3338.56 delta = 85.0898
[8] run: 1 lumi: 1 event: 8  vsize = 6372.39 deltaVsize = 0.0078125 rss = 3397.37 delta = 58.8047
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[10] run: 1 lumi: 1 event: 10  vsize = 6372.39 deltaVsize = 0 rss = 3408.45 delta = 11.0859
[9] run: 1 lumi: 1 event: 9  vsize = 6372.39 deltaVsize = 0 rss = 3405.04 delta = 7.66797
[8] run: 1 lumi: 1 event: 8  vsize = 6372.39 deltaVsize = 0.0078125 rss = 3397.37 delta = 58.8047
TimeReport> Time report complete in 281.325 seconds
 Time Summary: 
 - Min event:   2.63794
 - Max event:   17.0326
 - Avg event:   5.85748
 - Total loop:  122.119
 - Total init:  154.509
 - Total job:   281.325
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.0818873 ev/s
 CPU Summary: 
 - Total loop:  115.4
 - Total init:  156.349
 - Total extra: 0
 - Total job:   275.98
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

step 3 (WF 34834.0)

 MemoryReport> Peak virtual size 10830.3 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 9805.86 deltaVsize = 0 rss = 7058.96 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 10830.3 deltaVsize = 1024.49 rss = 7500.25 delta = 441.285
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[4] run: 1 lumi: 1 event: 4  vsize = 10830.3 deltaVsize = 0 rss = 7134.74 delta = -365.512
[3] run: 1 lumi: 1 event: 3  vsize = 10830.3 deltaVsize = 0 rss = 7416.19 delta = -84.0586
[2] run: 1 lumi: 1 event: 2  vsize = 10830.3 deltaVsize = 1024.49 rss = 7500.25 delta = 441.285
TimeReport> Time report complete in 2298.29 seconds
 Time Summary: 
 - Min event:   141.309
 - Max event:   332.148
 - Avg event:   186.191
 - Total loop:  2040.3
 - Total init:  252.703
 - Total job:   2298.29
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.00490124 ev/s
 CPU Summary: 
 - Total loop:  1494.69
 - Total init:  102.663
 - Total extra: 0
 - Total job:   1602.19
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

--------------------------------------------------------------------------------------

**new code**

step 3 (WF 34634.0)

MemoryReport> Peak virtual size 6374.41 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 6374.36 deltaVsize = 0 rss = 3375.52 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 6374.41 deltaVsize = 0.046875 rss = 3448.57 delta = 73.0586
[8] run: 1 lumi: 1 event: 8  vsize = 6374.41 deltaVsize = 0.0078125 rss = 3500.94 delta = 52.3633
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[10] run: 1 lumi: 1 event: 10  vsize = 6374.41 deltaVsize = 0 rss = 3509.6 delta = 8.66016
[9] run: 1 lumi: 1 event: 9  vsize = 6374.41 deltaVsize = 0 rss = 3513.95 delta = 13.0078
[8] run: 1 lumi: 1 event: 8  vsize = 6374.41 deltaVsize = 0.0078125 rss = 3500.94 delta = 52.3633
TimeReport> Time report complete in 219.515 seconds
 Time Summary: 
 - Min event:   2.40397
 - Max event:   16.8583
 - Avg event:   5.15732
 - Total loop:  106.134
 - Total init:  109.03
 - Total job:   219.515
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.0942203 ev/s
 CPU Summary: 
 - Total loop:  100.611
 - Total init:  135.833
 - Total extra: 0
 - Total job:   240.449
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

step 3 (WF 34834.0)

MemoryReport> Peak virtual size 10829.4 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 9804.88 deltaVsize = 0 rss = 6334.34 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 10829.4 deltaVsize = 1024.49 rss = 6906.25 delta = 571.91
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[4] run: 1 lumi: 1 event: 4  vsize = 10829.4 deltaVsize = 0 rss = 6600.7 delta = -305.551
[3] run: 1 lumi: 1 event: 3  vsize = 10829.4 deltaVsize = 0 rss = 6856.95 delta = -49.3008
[2] run: 1 lumi: 1 event: 2  vsize = 10829.4 deltaVsize = 1024.49 rss = 6906.25 delta = 571.91
TimeReport> Time report complete in 2217.3 seconds
 Time Summary: 
 - Min event:   143.84
 - Max event:   377.531
 - Avg event:   193.618
 - Total loop:  2092.58
 - Total init:  119.324
 - Total job:   2217.3
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.00477879 ev/s
 CPU Summary: 
 - Total loop:  1516.13
 - Total init:  96.4185
 - Total extra: 0
 - Total job:   1617.48
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

--------------------------------------------------------------------------------------

**DQM file size**
current:
1129.74 KiB      MTD/ETL
239.63 KiB       MTD/BTL

new:
1535.33 KiB      MTD/ETL
560.66 KiB       MTD/BTL

### CODE REVIEW

File: Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc [hunk 1]
```diff
@@ -489,6 +584,52 @@ void BtlLocalRecoValidation::bookHistograms(DQMStore::IBooker& ibook,
...
+  meCluRhoRes_ =
+      ibook.book1D("BtlCluRhoRes", "BTL cluster #rho resolution;#rho_{RECO}-#rho_{SIM} [cm]", 100, -0.5, 0.5);
+  meCluPhiRes_ =
+      ibook.book1D("BtlCluPhiRes", "BTL cluster #phi resolution;#phi_{RECO}-#phi_{SIM} [rad]", 100, -0.03, 0.03);
+  if (LocalPosDebug_) {
+    meCluXRes_ = ibook.book1D("BtlCluXRes", "BTL cluster X resolution;X_{RECO}-X_{SIM} [cm]", 100, -3.1, 3.1);
+    meCluYRes_ = ibook.book1D("BtlCluYRes", "BTL cluster Y resolution;Y_{RECO}-Y_{SIM} [cm]", 100, -3.1, 3.1);
+    meCluZRes_ = ibook.book1D("BtlCluZRes", "BTL cluster Z resolution;Z_{RECO}-Z_{SIM} [cm]", 100, -0.2, 0.2);
+  }
+  meCluYXLocal_ = ibook.book2D("BtlCluYXLocal",
```
Comment (fabiocos) [inline Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc:616]: @casarsa don't we want to keep these local position histograms as optional (with the flag)?

File: Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc [hunk 2]
```diff
@@ -814,6 +903,92 @@ void EtlLocalRecoValidation::bookHistograms(DQMStore::IBooker& ibook,
...
+                        "S");
+  meCluXRes_[0] = ibook.book1D("EtlCluXResZneg", "ETL cluster X resolution (-Z);X_{RECO}-X_{SIM} [cm]", 100, -0.1, 0.1);
+  meCluXRes_[1] = ibook.book1D("EtlCluXResZpos", "ETL cluster X resolution (+Z);X_{RECO}-X_{SIM} [cm]", 100, -0.1, 0.1);
+  meCluYRes_[0] = ibook.book1D("EtlCluYResZneg", "ETL cluster Y resolution (-Z);Y_{RECO}-Y_{SIM} [cm]", 100, -0.1, 0.1);
+  meCluYRes_[1] = ibook.book1D("EtlCluYResZpos", "ETL cluster Y resolution (+Z);Y_{RECO}-Y_{SIM} [cm]", 100, -0.1, 0.1);
+  meCluZRes_[0] =
+      ibook.book1D("EtlCluZResZneg", "ETL cluster Z resolution (-Z);Z_{RECO}-Z_{SIM} [cm]", 100, -0.003, 0.003);
+  meCluZRes_[1] =
+      ibook.book1D("EtlCluZResZpos", "ETL cluster Z resolution (+Z);Z_{RECO}-Z_{SIM} [cm]", 100, -0.003, 0.003);
+  meCluYXLocal_[0] = ibook.book2D("EtlCluYXLocalZneg",
```
Comment (fabiocos) [inline Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc:959]: @casarsa same here

**Comment by fabiocos** [comment]
@casarsa do you have the output plots to attach as documentation? In particular those showing the behaviour studied in the uncertainties tuning?

Do you always manage to find the simhits associated to the cluster? Do you have a check for this?

File: Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc (hunk 1 above)
Comment (casarsa) [reply Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc:616]: You are right. I move them into the if satetment

File: Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc (hunk 2 above)
Comment (casarsa) [reply Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc:959]: OK

File: Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc (hunk 1 above)
Comment (casarsa) [reply Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc:616]: and now the size of the MTD histograms is
1135.12 KiB      MTD/ETL
241.00 KiB       MTD/BTL

**Comment by casarsa** [comment]
> Do you always manage to find the simhits associated to the cluster? Do you have a check for this?

I've added a check. From a quick check: out of ~380 BTL reco hits and ~370 reco ETL I always have a match to a sim hit.

**Comment by fabiocos** [comment]
please test

@casarsa thanks for checking and the update

**Comment by jfernan2** [comment]
+1

**Comment by casarsa** [comment]
> @casarsa do you understand the pattern of the differences? Why no resolution plot and ETL plot? Didi you see them in the local test output?

Yes, I think I do:
BtlLongposPullvsEta: I changed the SIM hit position in the crystal from mm to cm and this affects the eta.

BtlCluPhi and BtlCluEta: I fixed the calculation of the cluster local position in the module. RectangularMTDTopology::pixelToModuleLocalPoint was used to transform coordinates that already were in the module frame.

BtlTPullvsEta: I've changed the number of bins for uniformity with other similar histograms.

**Comment by fabiocos** [comment]
@casarsa thanks, the new histograms are indeed available in the output

**Comment by qliphy** [comment]
+1

**Pull request merged**